types/datasource.d.ts

```diff
@@ -0,0 +1,169 @@
+// Node module: loopback-datasource-juggler
+// This file is licensed under the MIT License.
+// License text available at https://opensource.org/licenses/MIT
+
+import {Callback, PromiseOrVoid, Options, AnyObject} from './common';
+import {
+  ModelBuilder,
+  ModelBase,
+  ModelDefinition,
+  PropertyDefinition,
+} from './model';
+
+/**
+ * LoopBack models can manipulate data via the DataSource object.
+ * Attaching a `DataSource` to a `Model` adds instance methods and static methods to the `Model`.
+ *
+ * Define a data source to persist model data.
+ * To create a DataSource programmatically, call `createDataSource()` on the LoopBack object; for example:
+ * ```js
+ * var oracle = loopback.createDataSource({
+ *   connector: 'oracle',
+ *   host: '111.22.333.44',
+ *   database: 'MYDB',
+ *   username: 'username',
+ *   password: 'password'
+ * });
+ * ```
+ *
+ * All classes in single dataSource share same the connector type and
+ * one database connection.
+ *
+ * For example, the following creates a DataSource, and waits for a connection callback.
+ *
+ * ```
+ * var dataSource = new DataSource('mysql', { database: 'myapp_test' });
+ * dataSource.define(...);
+ * dataSource.on('connected', function () {
+ *     // work with database
+ * });
+ * ```
+ * @class DataSource
+ * @param {String} [name] Optional name for datasource.
+ * @options {Object} settings Database-specific settings to establish connection (settings depend on specific connector).
+ * The table below lists a typical set for a relational database.
+ * @property {String} connector Database connector to use.  For any supported connector, can be any of:
+ *
+ * - The connector module from `require(connectorName)`.
+ * - The full name of the connector module, such as 'loopback-connector-oracle'.
+ * - The short name of the connector module, such as 'oracle'.
+ * - A local module under `./connectors/` folder.
+ * @property {String} host Database server host name.
+ * @property {String} port Database server port number.
+ * @property {String} username Database user name.
+ * @property {String} password Database password.
+ * @property {String} database Name of the database to use.
+ * @property {Boolean} debug Display debugging information. Default is false.
+ *
+ * The constructor allows the following styles:
+ *
+ * 1. new DataSource(dataSourceName, settings). For example:
+ *   - new DataSource('myDataSource', {connector: 'memory'});
+ *   - new DataSource('myDataSource', {name: 'myDataSource', connector: 'memory'});
+ *   - new DataSource('myDataSource', {name: 'anotherDataSource', connector: 'memory'});
+ *
+ * 2. new DataSource(settings). For example:
+ *   - new DataSource({name: 'myDataSource', connector: 'memory'});
+ *   - new DataSource({connector: 'memory'});
+ *
+ * 3. new DataSource(connectorModule, settings). For example:
+ *   - new DataSource(connectorModule, {name: 'myDataSource})
+ *   - new DataSource(connectorModule)
+ */
```

Comment:
Is it a good idea to duplicate the jsdoc comments in typing files? I am concerned that the content in `.d.ts` files will quickly get out of sync with the original comments in `.js` files.

Reply:
When the package is used by other modules such as `@loopback/repository`, VSCode uses the tsdocs from `.d.ts` file, IIUC.

Reply:
I see, that makes sense.

It would be great to find a way how to automatically sync these jsdoc/tsdoc comments, but that's out of scope of this pull request.